### PR TITLE
[Feature] Add trainable frequency feature maps

### LINKED ISSUE #267: [Feature] Add trainable frequency feature map
Paper: https://scirate.com/arxiv/2309.03279

Example with old feature map structure:

```
def tf_feature_map(
    n_qubits: int,
    support: tuple[int, ...] = None,
    param: str = "phi",
    vparam: str = "omega",
    op: Type[Rotation] = RX,
    basis_set: str = "fourier",
) -> KronBlock:
    fparam = FeatureParameter(param)
    vparams = [VariationalParameter(f"{vparam}_{i}") for i in range(n_qubits)]

    if support is None:
        support = tuple(range(n_qubits))

    assert len(support) <= n_qubits, "Wrong qubit support supplied"

    if basis_set == "fourier":
        fm = kron(*[op(qubit, vparams[qubit] * fparam) for qubit in support])
    elif basis_set == "chebyshev":
        fm = kron(
            *[op(qubit, 2 * sympy.acos(vparams[qubit] * fparam)) for qubit in support]
        )
    elif basis_set == "tower":
        fm = kron(
            *[
                op(qubit, (i + 1) * 2 * sympy.acos(vparams[qubit] * fparam))
                for i, qubit in enumerate(support)
            ]
        )
    else:
        raise NotImplementedError(f"Feature map {basis_set} not implemented")
    fm.tag = "FM"
    return fm
```

### PR DESCRIPTION
**Pull request by jpmoutinho**
Closes #267

### CODE REVIEW

File: docs/qml/qml_constructors.md [hunk 1]
```diff
@@ -85,6 +83,27 @@ from qadence.draw import html_string # markdown-exec: hide
...
+    fm_type=BasisSet.FOURIER,
+    reupload_scaling=ReuploadScaling.EXP,
+    param_prefix = "w",
+)
+
+from qadence.draw import html_string # markdown-exec: hide
+print(html_string(fm_trainable)) # markdown-exec: hide
+```
+
+Note that for the Fourier feature map the encoding function is simply $f(x)=x$. For other cases, like the Chebyshev `acos()` encoding,
```
Comment (Roland-djee) [inline docs/qml/qml_constructors.md:103]:
```suggestion
Note that for the Fourier feature map, the encoding function is simply $f(x)=x$. For other cases, like the Chebyshev `acos()` encoding,
```

File: qadence/constructors/feature_maps.py [hunk 2]
```diff
@@ -76,21 +82,28 @@ def fm_parameter(
...
-    elif inspect.isclass(fm_type) and issubclass(fm_type, Function):
-        transformed_feature = fm_type(scaled_fparam)
+        transform_func = acos
+    elif callable(fm_type):
+        transform_func = fm_type
     else:
         raise NotImplementedError(
             f"Feature map type {fm_type} not implemented. Choose an item from the BasisSet "
-            f"enum: {[bs.name for bs in BasisSet]}, or your own sympy.Function to wrap "
+            f"enum: {[bs.name for bs in BasisSet]}, or your own sympy function to wrap "
```
Comment (Roland-djee) [inline qadence/constructors/feature_maps.py:102]:
```suggestion
            f"enum: {BasisSet.list()}, or a custom defined sympy function to wrap "
```

File: qadence/constructors/feature_maps.py [hunk 3]
```diff
@@ -76,21 +82,28 @@ def fm_parameter(
     else:
         scaled_fparam = scaling * fparam + shift
 
+    return scaled_fparam
+
+
+def fm_parameter_func(fm_type: BasisSet | Callable | str) -> type[Function]:
```
Comment (Roland-djee) [inline qadence/constructors/feature_maps.py:88]:
```suggestion
def fm_parameter_func(fm_type: BasisSet | Callable | str) -> Callable:
```

File: qadence/constructors/feature_maps.py [hunk 4]
```diff
@@ -76,21 +82,28 @@ def fm_parameter(
...
     else:
         raise NotImplementedError(
             f"Feature map type {fm_type} not implemented. Choose an item from the BasisSet "
-            f"enum: {[bs.name for bs in BasisSet]}, or your own sympy.Function to wrap "
+            f"enum: {[bs.name for bs in BasisSet]}, or your own sympy function to wrap "
             "the given feature parameter with."
         )
 
-    return transformed_feature
+    return transform_func  # type: ignore [return-value]
```
Comment (Roland-djee) [inline qadence/constructors/feature_maps.py:106]: With the return type suggestion, could the ignore be removed ?

**Review by Roland-djee — APPROVED** [review]
Thanks @jpmoutinho. Minor suggestions from me. Question: is it worth testing this ?

File: qadence/constructors/feature_maps.py (hunk 4 above)
Comment (jpmoutinho) [reply qadence/constructors/feature_maps.py]: Yes indeed.

**Comment by jpmoutinho** [comment]
> Thanks @jpmoutinho. Minor suggestions from me. Question: is it worth testing this ?

Thanks @Roland-djee. I tested the addition of the trainable parameter in the creation and running of the feature map for some given input values so it should be ok in terms of functionality, but I'm not sure there anything we test in terms of "correctness" here.

**Pull request merged**